cmake/modules/dts.cmake

```diff
@@ -106,6 +106,7 @@ set(DTS_DEPS                    ${PROJECT_BINARY_DIR}/zephyr.dts.d)
 # This generates DT information needed by the Kconfig APIs.
 set(GEN_DRIVER_KCONFIG_SCRIPT   ${DT_SCRIPTS}/gen_driver_kconfig_dts.py)
 # Generated Kconfig symbols go here.
+set_ifndef(KCONFIG_BINARY_DIR   ${CMAKE_BINARY_DIR}/Kconfig)
```

Comment:
I'm not sure if setting a default value for a variable like this is the right policy. Certain build system-wide variables should probably be centralized, rather than letting each module set its own default value. That could lead to its own brand of subtle bugs in the long term.

There's also this comment further up, which already suggests that `KCONFIG_BINARY_DIR` is expected to be defined before this module is loaded:

```
# Required variables:
# - BINARY_DIR_INCLUDE_GENERATED: where to put generated include files
# - KCONFIG_BINARY_DIR: where to put generated Kconfig files
```

Would it be better to do the following instead? (in a separate PR)
- Add `include(generated_file_directories)` and `include(zephyr_module)`.
- Check if every required variable is defined; raise an error if not.

Reply:
> I'm not sure if setting a default value for a variable like this is the right policy.

Correct, it's not ideal, but improving is part of a longer term goal, as also mentioned here:
https://github.com/zephyrproject-rtos/zephyr/pull/55293#discussion_r1122798737

> There's also this comment further up, which already suggests that KCONFIG_BINARY_DIR is expected to be defined before this module is loaded:

which is also not ideal, generally speaking a module should be as self-contained as possible with proper handling of required and optional value, and sane defaults for parts which are not core for the module.

In this case, creation of Kconfig file should not be core to a dts module, cause one should be able to load just dts.cmake for devicetree creation without Kconfig if there is a need. (such as `cmake-package-helper filtering).

Therefore creation of a default value is better than to fail if not set.
But ideally the Kconfig files should probably not be created if the variable has not been defined.

> Would it be better to do the following instead? (in a separate PR)

Yes, I think so, I believe this might be left-over from #55616 where iirc it is actually needed.
Will try to move it out of this PR.

Reply:
Removed commits and code related to KCONFIG_BINARY_DIR from sysbuild.